### PR TITLE
feat(dracut): support parallel execution with --parallel

### PR DESCRIPTION
**Pull request by mwilck**
Add an option `--parallel` which can currently be used with
`--regenerate-all`. With `--regenerate-all --parallel`, dracut will be
run in parallel for all kernels found.

Also introduce a config file equivalent setting: `parallel="yes"`.

In my testing I've found a linear speed up in execution time.

## Changes

 * add option `--regenerate-parallel`

## Checklist
- [x] I have tested it locally
- [x] I have reviewed and updated any documentation if relevant
- [ ] I am providing new code and test(s) for it

### CODE REVIEW

**Comment by aafeijoo-suse** [comment]
I like the idea, we gain speed on systems with multiple kernels, although the log gets messy.
By the way, you need to fix some formatting issues, `shfmt -s -w dracut.sh` will do the job (see https://github.com/dracutdevs/dracut/runs/6814635981?check_suite_focus=true)

**Comment by mwilck** [comment]
> I like the idea, we gain speed on systems with multiple kernels, although the log gets messy. By the way, you need to fix some formatting issues, `shfmt -s -w dracut.sh` will do the job (see https://github.com/dracutdevs/dracut/runs/6814635981?check_suite_focus=true)

Fixed.

**Comment by johannbg** [comment]
I'm not particular happy about --regenerate-all[=yes] and --regenerate-all=parallel since it will just end up confusing users, that said is there any argument against simply running things in parallel besides logs getting messy when it's run? ( which arguably is to be expected from parallel executions meaning it's a none issue from my pov )

**Comment by mwilck** [comment]
If you prefer doing this unconditionally, fine with me. I opted for `--regenerate-all=parallel` because it doesn't change the default behavior (`--regenerate-all` still works as usual), and I wanted to avoid introducing a new, separate option for this.

If you fear that people get confused (why would they?), perhaps I should just leave the feature undocumented (i.e. undo the man page change)?

**Comment by johannbg** [comment]
> If you prefer doing this unconditionally, fine with me. I opted for `--regenerate-all=parallel` because it doesn't change the default behavior (`--regenerate-all` still works as usual), and I wanted to avoid introducing a new, separate option for this.
> 
> If you fear that people get confused (why would they?), 

Because regenerate-all already implies you are going to regenerate all initramfs images and now you have added an "=yes" switch into the mix which obviously implies that there should be a "=no" to accommodate it, which in turn should do something different so users might for example expect that it regenerates not all but "some" initramfs images or regenerate just one initramfs image etc.

The same applies to  --regenerate-all=parallel then users expect being able to configure an swith opposide to that etc.

Having a "run-parallel" option ( and parallel=on/off in conf file ) is a better solution compared to having --regenerate-all=parallel to trigger this or --regenerate-all=parallel-off to disable it if set to be the default from my pov but ymmv.

**Comment by aafeijoo-suse** [comment]
> ... besides logs getting messy when it's run? ( which arguably is to be expected from parallel executions meaning it's a none issue from my pov )

I agree, I don't see it as a problem, plus we can document that this will happen during parallel build.

> Having a "run-parallel" option ( and parallel=on/off in conf file ) is a better solution compared to having --regenerate-all=parallel to trigger this or --regenerate-all=parallel-off to disable it if set to be the default from my pov but ymmv.

I'd also vote to have a separate option for this, and keep the meaning of `--regenerate-all` as it is now, but that's also just a personal preference.

**Comment by mwilck** [comment]
I've changed the option syntax.

**Comment by sinyb** [comment]
May I suggest that instead of adding sub-options to "--regenerate-all", we introduce new option (-p is currently unused, or --parallel), that would have no effect for single kernel/initramfs?
That way it would be trivial to add "parallel=yes" to default config file and have wanted behavior with little effort. 
Programs that run dracut "behind the scene" (like zypper) would not have to know anything about the change...

**Comment by sinyb** [comment]
> I've changed the option syntax.

```
/etc/dracut.conf
regenerate_parallel="yes"
```
has no effect with "dracut --regenerate-all  --force", while
`regenerate-parallel="yes"`
causes error "/etc/dracut.conf: line 5: regenerate-parallel=yes: command not found"

**Comment by mwilck** [comment]
> ```
> /etc/dracut.conf
> regenerate_parallel="yes"
> ```
> 
> has no effect with "dracut --regenerate-all --force", while `regenerate-parallel="yes"` causes error "/etc/dracut.conf: line 5: regenerate-parallel=yes: command not found"

Hm, this is tricky. `--regenerate-all` (and thus, `regenerate_parallel`) is evaluated before reading the config file. Therefore with this patch, setting `regenerate_parallel` can't be set in the config file.

**Comment by mwilck** [comment]
I pushed another commit that allows using `regenerate_all="yes"` in the conf file.

**Comment by johannbg** [comment]
I think it's better to just call this parallel which allows for it to be potentially extended in the future as opposed to call it --regenerate-parallel and just document it can only be used along with  --regenerate-all at the moment and add an short switch of -p to it as well also remember to update the shell completion.

**Comment by mwilck** [comment]
> I think it's better to just call this parallel which allows for it to be potentially extended in the future as opposed to call it --regenerate-parallel and just document it can only be used along with --regenerate-all at the moment and add an short switch of -p to it as well also remember to update the shell completion.

You're not concerned that `--parallel` might confuse users by raising unrealistic expectations?

**Comment by sinyb** [comment]
> > I think it's better to just call this parallel which allows for it to be potentially extended in the future as opposed to call it --regenerate-parallel and just document it can only be used along with --regenerate-all at the moment and add an short switch of -p to it as well also remember to update the shell completion.
> 
> You're not concerned that `--parallel` might confuse users by raising unrealistic expectations?

You mean like using parallel versions of zstd/xz/bzip2/gz/whatever in addition to creating multiple inirds in parallel?

I wouldn't be too concerned. Whoever comes that far to use --parallel has probably read man page or at least --help, where its behavior is properly documented...

Speaking of which, why not use parallel compression too? :)
Is that anything else that can be parallelized?

**Comment by mwilck** [comment]
> Speaking of which, why not use parallel compression too? :) 

Yes, it's used with `compress=zstd` (`zstd -15 -q -T0`). But only zstd yields reproducible results with multithreading, therefore it's not enabled for other algorithms.

**Comment by sinyb** [comment]
> Yes, it's used with `compress=zstd` (`zstd -15 -q -T0`). But only zstd yields reproducible results with multithreading, therefore it's not enabled for other algorithms.

Nice to know, thanks! 
It seems that it is already default on my Opensuse Tumbleweed, I get the same results with or without " --compress=`zstd -15 -q -T0` ":
time dracut --regenerate-all -f
real    0m6.960s
user    0m38.493s
sys     0m12.846s
(for 6 kernels, I have added regenerate_parallel="yes" to .conf)

**Comment by mwilck** [comment]
OK, renamed the option to `-p | --parallel` as requested.

File: dracut.sh
Comment (johannbg) [inline dracut.sh:959]: You forgot to change `--regenerate-parallel`  to `-p`, `--parallel` here

**Review by johannbg — CHANGES_REQUESTED** [review]
Change your commit message
from `feat: support parallel execution with --parallel`
to 
`feat(dracut): support parallel execution with --parallel`

Install and run shfmt -s -w dracut.sh to fix the formatting issues with your changes
and update `--regenerate-parallel` to `-p` `--parallel` where you forgot it.

File: dracut.sh
Comment (mwilck) [reply dracut.sh:959]: right, thanks. But as `--parallel` is supposed to be independent of `--regenerate-all`, maybe I should pass the flag down to the child processes instead? (currently it won't have any effect as `--regenerate-all` is removed).

Comment (mwilck) [reply dracut.sh]: re-pushed; still removing `--parallel`. Tell me what you prefer.

Comment (johannbg) [reply dracut.sh]: Let's keep this as is

**Review by johannbg — APPROVED** [review]
LGTM once the test have run I'll merge this.

**Comment by mwilck** [comment]
@aafeijoo-suse, any chance to get this in TW soon?

**Comment by aafeijoo-suse** [comment]
> @aafeijoo-suse, any chance to get this in TW soon?

Yes, the update to 057 is on the devel project and will be submitted to factory soon.

**Pull request merged**
